text/083-boost-stdlib-dev/text.md

```diff
@@ -0,0 +1,786 @@
+  * <details><summary>contains:</summary>
+    theories/derive/Derive.v</details>
+* coq-roots-rtauto
+  * The Coq standard library, rtauto
+  * Namespace `Coq.rtauto`
+  * 2 files, 0.8 kloc
+  * <details><summary>contains:</summary> rtauto/Bintree.v,
+    rtauto/Rtauto.v</details>
+* coq-roots-wellfounded
+  * The Coq standard library, wellfounded
+  * Namespace `Coq.wellfounded`
+  * 9 files, 0.9 kloc
+  * <details><summary>contains:</summary> Wellfounded/Inclusion.v,
+    Wellfounded/Wellfounded.v, Wellfounded/Union.v,
+    Wellfounded/Transitive_Closure.v, Wellfounded/Well_Ordering.v,
+    Wellfounded/Inverse_Image.v, Wellfounded/Disjoint_Union.v,
+    Wellfounded/Lexicographic_Product.v,
+    Wellfounded/Lexicographic_Exponentiation.v</details>
+* coq-roots
+  * The Coq standard library
+  * A metapackage for all coq-roots-* packages constituting
+    the historical Coq standard library, that used to be shipped with Coq.
+  * 508 files, 183.4 kloc
+
+theories/ssr and theories/ssrmatching should go back into mathcomp
+(either in coq-mathcomp-ssreflect or in a new coq-mathcomp-ssr
+containing only that).
+
+### Documentation
+
+Need to document the above split in packages (could help structure the doc).
+
+## Maintainers
+
+We should do an open call for maintainers. Current maintainers (stdlib
+team from Coq on github) will become roots maintainer, they can of
+course opt out at any time.
+
+# Drawbacks
+
+The stdlib will no longer come bundled with Coq. It will however be
+just as easy to install via the usual package managers. More notably
+it will remain an integral part of the Coq Platform.
+
+# Alternatives
+
+The technical changes proposed here can stand on their own merits, and can be adopted without ratifying the perspective presented for motivating them. Further, each of the three splits discussed under [Packages](https://github.com/proux01/ceps/blob/boost-stdlib-dev/text/083-boost-stdlib-dev/text.md#packages) could be performed without the other ones, with or without the broader vision that led to them.
+
+1. The logical structure of the library could be split into packages, enforced by the build system, without user-visible structural changes or renaming of stdlib.
+2. The standandard library could be moved to its own repository without changing its user-facing relationship with Coq and non-stdlib packaes.
+3. Specific parts of the standard library could be spun out to their own repositories without changing the rest.
+
+Out of these, 1. has the advantage that it retains most of the technical refactoring done for this CEP, and achieves some of the technical goals listed here such as allowing users to `Require` only a known subset, while minimizing disruption to users in the face of multiple possible futures for the standard library more generally. In particular, keeping the boundaries between subparts of the standard library internal to Coq as opposed to exposing them to users would simplify later changing them.
```

Comment:
[quoting precisely because the line is way too long to attach comments to a specific part]

> In particular, keeping the boundaries between subparts of the standard library internal to Coq as opposed to exposing them to users would simplify later changing them.

That's a good argument against exposing the package split (point 2. in https://github.com/proux01/ceps/blob/boost-stdlib-dev/text/083-boost-stdlib-dev/text.md#packages ).

My feeling is that you are strongly opposed to the (somewhat orthogonal) dualrepo choice. What are the concrete \
arguments behind that position, I can see

* monorepo is a way to strongly guarantee no additional dependency can be added

are there any other?

Reply:
The key advantage of the monorepo is the one that you already listed in the CEP: it drastically simplifies the logistics of making a change that affects multiple parts (compared to these parts being in separate repositories). A special case of this is that a monorepo makes it easy to ensure internal compatibility by testing every PR in a single CI. Making changes that (conceptually simultaneously) touch coqc, stdlib, and CI developments is a key part of my vision for how to evolve this ecosystem, so I am indeed setting a higher bar for splitting the repo.

While tooling for different workflows within a monorepo can go a long way, I am yet to see a concept for smooth and reliable coordination of changes across different repositories. Overlays as in Coq are very useful given the multirepo reality of Coq CI but they have merge races and pull-request toil (and delays), pinned submodules as in fiat-crypto act as a monorepo with an extra step when everything is moving smoothly and cause an effective freeze of the subrepo code otherwise, maintaining a stable interface across versions like Coq currently strives for is already very limiting on what kind of changes Coq developers can make and I worry it would be even more limiting if the stdlib plugins became a part of that standard interface. Maintaining compatibility in a multirepo environment is also a constant chore in BoringSSL which I sometimes contribute to: there is a porting-duty rotation between the maintainers, and big changes are still tested with supported users before merging anywhere. Basically, every arrangement where a library actually supports its users requires adapting these users to changes at some point, and being able to do this atomically in a single commit in the single repository is the least painful mechanism I have seen, by far.

For a monorepo to make sense, the maintainers of that repository have to have working baseline agreement about what to do with it and mutual trust that only changes that are in line with this agreement are merged. IMO the required consistency bar for is rather low, but it is not zero -- this is why I focus on process questions in #86.  Agreeing to disagree can still work well, and it is not hard to imagine e.g. separate agreed-upon and perhaps even coqbot-enforced rules for different CODEOWNERS. On the other hand, treating some piece as an external dependency with an independent release schedule and compatibility story seems to inherently come with human toil for each coupled change, so I would explore this option only when synchronized development is known to be substantially at odds with the goals of one project or another. I get the sense that you probably do have a vision for stdlib that is at odds with it remaining a part of Coq, but I have only a vague sense for what it is, so it cannot convince me to give up the monorepo.

And as a user of coqc+plugins+stdlib, I count them as "one dependency" exactly because I trust the maintainers of these parts to coordinate effectively enough between themselves that I never have to think about compat between them. (Taking this view to a silly extreme, Debian is "one dependency" too, but it is not the same dependency as Coq, and trying to work with both of them is indeed constraining for fiat-crypto).

Reply:
Thanks for the explanation. So the root of our disagreement seems to be a strong "philosophical" opinion about monorepo vs. distributed developments with version management. Some people tend to think that repository boundaries should reflect human team boundaries, whereas you'd prefer something wider.

[off topic but maybe interesting]
About the CI of Coq: it seems we also have a quite different approach of it. You seem to see it as some approximation of your monorepo dream. My dream is rather an open world where everyone can maintain its own Coq development without any need to be in the CI, only by easy adaptation to each new Coq / other-dependency release, following its changelog and warning / errors.

You seem to expect a lot from the CI, whereas IMHO, the CI is only a kind of expansive test suite and developers of developments in the CI should only expect:
* less Coq bugs on their dev since CI should catch them early;
* getting involved early in discussions for breaking changes when such changes are needed.

in exchange of having to constantly remain compatible with master.
(Plugins are a particular case where we commit to provide overlays, due to the unstability of the API.)
[/off topic but maybe intresting]

More precised answer to detailled points:

> Overlays as in Coq [...] have merge races

This is theoretically true. However, in my experience, the few merge races I did experience in Coq where never related to overlays. So I wonder how practical the issue is.

> and pull-request toil (and delays)

Right, although again the comment mostly make sense when both sides of the overlay are done by the same "team". Those costs are also a good way to get an idea of how, and at which cost, external devs can be adapted.

> pinned submodules as in fiat-crypto act as a monorepo

Ironically enough, as a Coq developper, my feeling is that those are responsible for a large part of the "overlay toil". Again, only my personal experience, not a fact.

> I get the sense that you probably do have a vision for stdlib that is at odds with it remaining a part of Coq

If by "part of Coq" you mean "same repo", then the main technical point of this CEP is the dualrepo choice, so yes of course. If you mean something else, please clarify.

In any case, it seems we mostly share the same vision of an ideal stdlib. It also seems that the current stdlib should remain usable by all its current users (whether in CI or not) without crazy amounts of effort. Whether the two goals are compatible remains unclear to me.

Reply:
> However, in my experience, the few merge races I did experience in Coq where never related to overlays.

Okay, maybe it's not specifically overlays at fault here. But the loose coupliing of CI developments is definitely a major cause behind CI failures: 4 out of 5 last non-cancellation failures look like an atomically-merge-when-tests-pass policy would have prevented them: https://github.com/coq/coq/runs/23898721987, https://github.com/coq/coq/runs/23874454446, https://github.com/coq/coq/runs/23874454446, https://github.com/coq/coq/runs/23874782784,https://github.com/coq/coq/runs/23874833864

> Those costs are also a good way to get an idea of how, and at which cost, external devs can be adapted.

For stdlib changes, IME the costs of actually adapting the developments are a tiny fraction of the coordination overhead of getting the right person to look at the change and click merge because it is too boring to look at. I think it has happened to me a grand total of once that feedback from a adaptation PR led to a change in the stdlib work.

> Ironically enough, as a Coq developper, my feeling is that those are responsible for a large part of the "overlay toil". Again, only my personal experience, not a fact.

I mean, I'm not saying the submodules are great, just the alternative would be fiat-crypto being red as often as other large multirepo developments. And the particularly slow and painful cases I've witnessed were exactly because some non-coqdev-related codebase-synchronization issues needed to be resolved before the coqdev patch could be accepted.

> Whether the two goals are compatible remains unclear to me.

Which goals are you referring to?

Reply:
> 4 out of 5 last non-cancellation failures [...]

Those are not "real failures" but the consequence of synchronous overlay merge being indeed non atomic. Synchronous overlays are only a module thing. Since modules would remain in Coq repo, the issue wouldn't affect the library repo. So we	can either consider the	issue orthogonal to the current CEP or even the CEP being an improvement here, since it releases the library from all issues linked to synchronous overlays.

> I'm not saying the submodules are great, just the alternative would be fiat-crypto being red

[off topic]
I really have the feeling that coq-util / bedrock2 / rupicola / fiat-crypto / ... should live in a single repo.
From my outsider point of view, their team seem to have a pretty strong intersection.
[/off topic]

> > Whether the two goals are compatible remains unclear to me.

> Which goals are you referring to?

The ones from the previous sentence:
* evolve the current stdlib to make it truly "standard" / state of the art
* not dramatically breaking every uses

Anyway, that's definitely beyond the scope of the current CEP.

Reply:
I personally agree about fiat-crypto and its dependencies.

Reply:
> Synchronous overlays are only a module thing

I agree currently the two are associated, but are you sure this is fundamental? The sense I get is that the issues currently affecting ML plugins are fundamentally about interfaces that have not been designed for general consumption ending up at a repository boundary, regardless of implementation language. And the interfaces between ML and Gallina/Ltac code for stdlib plugins seem very much like that for me. Another experience confirming this has been maintaining a [copy of non-ML parts of nsatz](https://github.com/mit-plv/fiat-crypto/commits/5563f0925ec0187b9927dbc21b1eab76f15ac7f3/src/Algebra/Nsatz.v) in fiat-crypto.

Reply:
I'm not seeing anything interesting in that commit history, what am I missing?

Reply:
Yeah, the history doesn't really tell a story on its own, and I don't remember well enough to recount it play-by-play. But my takeaway was that a change (making nsatz good to use for custom fields) that would have one commit (or perhaps two for external compat) in a monorepo dragged out over years. This is not purely a technical point on the monorepo/dualrepo choice, as another important factor was slowness and uncertainty of contributing to Coq regardless of synchronization, but nevertheless, it is a negative experience with maintaining Ltac code that wraps an ML plugin in Coq.

Reply:
I'm not sure what you mean by fundamental. Synchronous overlays are needed for ML code because the ML API is unstable.

In the context of this PR, maybe we should check (both in Coq and the library) that it is properly documented that synchronous overlays are only acceptable for ML code, not for anything else (meaning synchronous overlays would be unacceptable in the library repo).

[off topic]
I agree that due to this, maintaining a ML plugin is bound to be, to some extent, an uncomfortable experience, forcing you to either be compatible with a reduced set of Coq versions or have some mechanism for conditional compilation of multiple plugin versions.
[/off topic]

Reply:
> Synchronous overlays are needed for ML code because the ML API is unstable

> meaning synchronous overlays would be unacceptable in the library repo

I mean that the the Ltac/Gallina APIs for stdlib plugins are no better than the ML ones (it's not about the language!), so freezing them by leaving them at the repository boundary seems unwise.